mentat/include_files.py

```diff
@@ -44,104 +23,225 @@ def is_file_text_encoded(abs_path: Path):
-        elif path.is_dir():
-            nonignored_files = set(
-                map(
-                    lambda f: Path(os.path.realpath(path / f)),
-                    get_non_gitignored_files(path),
-                )
-            )
+def validate_and_format_path(
+    path: Path | str, cwd: Path, check_for_text: bool = True
+) -> Path:
+    """Validate and format a path.
+
+    Args:
+        `path` - A file path, file interval path, directory path, or a glob pattern
+        `check_for_text` - Check if the file can be opened. Default to True
 
-            file_paths_from_dirs.update(
-                filter(
-                    lambda f: (not check_for_text) or is_file_text_encoded(f),
-                    nonignored_files,
-                )
+    Return:
+        An absolute path
+    """
+    path = Path(path)
+
+    # Get absolute path
+    if path.is_absolute():
+        abs_path = path
+    else:
+        abs_path = cwd.joinpath(path)
+
+    # Resolve path (remove any '..')
+    abs_path = abs_path.resolve()
+
+    # Validate different path types
+    # File
+    if abs_path.is_file():
+        if not abs_path.exists():
+            raise PathValidationError(f"File {abs_path} does not exist")
+        if check_for_text and not is_file_text_encoded(abs_path):
+            raise PathValidationError(f"Unable to read file {abs_path}")
+    # File interval
+    elif len(str(abs_path).rsplit(":", 1)) > 1:
+        _interval_path, interval_str = str(abs_path).rsplit(":", 1)
+        interval_path = Path(_interval_path)
+        if not interval_path.exists():
+            raise PathValidationError(f"File {interval_path} does not exist")
+        if check_for_text and not is_file_text_encoded(interval_path):
+            raise PathValidationError(f"Unable to read file {interval_path}")
+        intervals = parse_intervals(interval_str)
+        if len(intervals) == 0:
+            raise PathValidationError(
+                f"Unable to parse intervals for path {interval_path}"
             )
+    # Directory
+    elif abs_path.is_dir():
+        if not abs_path.exists():
+            raise PathValidationError(f"Directory {abs_path} does not exist")
+    # Glob pattern
+    elif re.search(r"[\*\?\[\]]", str(path)):
+        pass
+    else:
+        raise PathValidationError(f"Unable to validate path {path}")
 
-    files_from_dirs = set(CodeFeature(path.resolve()) for path in file_paths_from_dirs)
-    return files_direct, files_from_dirs, invalid_paths
+    return abs_path
 
 
-def get_ignore_files(ignore_paths: list[Path]) -> set[Path]:
-    """Returns a set of files to ignore from a list of ignore paths."""
+def match_path_with_patterns(path: Path, patterns: Set[str]) -> bool:
+    """Check if the given absolute path matches any of the patterns.
 
-    ignore_paths, _ = expand_paths(ignore_paths)
+    TODO: enforce valid glob patterns? Right now we allow glob-like
+    patterns (the one's that .gitignore uses). This feels weird imo.
```

Comment:
Thoughts on this?

Reply:
I don't think we need this function at all; why aren't we just using glob.glob like we were before? Just match all non-gitignored files in the directory, and then subtract all globbed files.

Reply:
https://github.com/AbanteAI/mentat/pull/286#issuecomment-1821996841

Reply:
Is this a problem though? This adds a lot of complexity and room for error compared to what we were doing before for what seems like not much benefit